keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1440 @@
+* Contain an X.509 PodIdentity extension.
+
+These certificates are translated to an authenticator.Response that has:
+* `User.Name` set to
+  `system:serviceaccount:{PodIdentity.Namespace}:{PodIdentity.ServiceAccountName}`.
+* `User.UID` set to `PodIdentity.ServiceAccountUID`.
+* `User.Extra["authentication.kubernetes.io/pod-name"]` set to
+  `{PodIdentity.PodName}`
+* `User.Extra["authentication.kubernetes.io/pod-uid"]` set to
+  `{PodIdentity.PodUID}`
+* `User.Extra["authentication.kubernetes.io/node-name"]` set to
+  `{PodIdentity.NodeName}`
+  * `User.Extra["authentication.kubernetes.io/node-uid"]` set to
+  `{PodIdentity.NodeUID}`
+
+The net effect is that authenticating via a PodIdentity-bearing cert has the
+same object-bound behavior as authenticating via a bound service account token.
+
+During Alpha, this feature will be gated by the
+`PodCertificateKubeApiserverClientAcceptance` feature gate.
+
+#### API Server Pod Client Certificate Approver / Signer
+
+The kube-controller-manager is extended to support a new certificate signer,
+`kubernetes.io/kube-apiserver-client-pod`.
+
+The controller for this signer name automatically issues certificates for
+PodCertificateRequests that are addressed to it.  It performs no additional
+validation on PodCertificateRequests beyond the validation enforced by
+kube-apiserver.
+
+The certificates it issues are meant to be understood by kube-apiserver as
+described above.  They:
+* Have a CommonName set to strong random text.  (This maintains flexibility, in
+  case we wish to include a meaningful common name in the future.)
+* Have a PodIdentity extension (non-critical) that contains the information from
+  `PodCertificateRequest.Spec`.
+* Key usages consistent with the certificate being used as a TLS client
+  certificate or server certificate.
+* It uses a CA specified by a new pair of private key / root certificate flags
+  (although implementations are free to set this flag to point at the CA as
+  another signer).
+* It issues certificates that are valid for 24 hours, and tells Kubelet to begin
+  refreshing them at 12 hours.
+
+During Alpha, this feature is gated by the
+`PodCertificateKubeApiserverClientIssuance` feature gate.
+
+Because kube-apiserver will interpret the PodIdentity extension as expressing an
+object binding between the presented certificate and the pod named in the
+extension, once a pod is no longer running, all certificates that were issued to
+it will become invalid.  This, along with the short lifetime of the
+certificates, removes the need for a more permanent certificate revocation
+mechanism.
+
+#### Client-go InClusterConfig Enhancements
+
+If client-go `InClusterConfig()` detects a credential bundle file at
+`/var/run/secrets/kubernetes.io/serviceaccount/credentialbundle.pem`, it will
+configure itself to use the private key and client certificate chain from the
+bundle, periodically reloading them to catch when they are rotated.
```

Comment:
Hmmm, maybe for alpha we should make it be a separate top-level function so that you have to explicitly opt-in?  Like `rest.InClusterPodCertificateConfig()`?

Reply:
I am not sure of the right approach, but I do want to give folks the ability to not leak go routines.

Reply:
I've updated the kep with this approach.  In the future we can tackle trying to use these credentials by default.